wrapper/src/main/java/software/amazon/jdbc/plugin/cache/CacheMonitor.java

```diff
@@ -0,0 +1,917 @@
+            ClusterHealthState cluster = clusterEntry.getValue();
+            HealthState clusterState = cluster.getClusterHealthState();
+
+            // Skip monitoring for HEALTHY clusters (zero overhead)
+            if (clusterState == HealthState.HEALTHY && !healthCheckInHealthyState) {
+              continue;
+            }
+
+            hasActiveMonitoring = true;
+
+            // Check RW endpoint
+            boolean rwPingSuccess = executePing(cluster, true);
+            handlePingResult(cluster, true, rwPingSuccess);
+
+            // Check RO endpoint if different from RW
+            if (cluster.roEndpoint != null && !cluster.roEndpoint.equals(cluster.rwEndpoint)) {
+              boolean roPingSuccess = executePing(cluster, false);
+              handlePingResult(cluster, false, roPingSuccess);
+            }
+          }
+
+          // Calculate iteration duration and sleep
+          long iterationDuration = System.currentTimeMillis() - iterationStart;
+          long targetInterval = hasActiveMonitoring
+              ? TimeUnit.SECONDS.toMillis(CACHE_HEALTH_CHECK_INTERVAL)
+              : THREAD_SLEEP_WHEN_INACTIVE_MILLIS;
+          long sleepMillis = Math.max(0, targetInterval - iterationDuration);
+          sleep(sleepMillis);
+        } catch (final InterruptedException intEx) {
+          throw intEx;
+        } catch (final Exception ex) {
+          LOGGER.warning("Exception during cache monitoring: " + ex.getMessage());
+        }
+      }
+    } catch (final InterruptedException intEx) {
+      LOGGER.warning("Cache monitor thread interrupted");
+    } finally {
+      this.stopped = true;
+      LOGGER.info("Cache monitor thread stopped");
+    }
+  }
+
+  /**
+   * Executes a ping operation to check cache endpoint health.
+   */
+  private boolean executePing(ClusterHealthState cluster, boolean isRw) {
+    String endpoint = isRw ? cluster.rwEndpoint : cluster.roEndpoint;
+    long lastPingTimestamp = isRw ? cluster.rwLastPingTimestamp : cluster.roLastPingTimestamp;
+
+    try {
+      long currentTime = System.currentTimeMillis();
+      long timeSinceLastPing = lastPingTimestamp == 0 ? 0 : currentTime - lastPingTimestamp;
+
+      String humanReadableTime = formatTimestamp(currentTime);
+
+      // Log ping start
+      if (lastPingTimestamp == 0) {
+        LOGGER.info(String.format(
+            "[PING_START] Endpoint: %s (%s) | Time: %s | First ping for this endpoint",
+            endpoint, isRw ? "RW" : "RO", humanReadableTime));
+      } else {
+        LOGGER.info(String.format(
+            "[PING_START] Endpoint: %s (%s) | Time: %s | Interval since last: %dms (%.2fs)",
+            endpoint, isRw ? "RW" : "RO", humanReadableTime, timeSinceLastPing, timeSinceLastPing / 1000.0));
+      }
+
+      long pingStartTime = System.currentTimeMillis();
+      boolean result = ping(cluster, isRw);
+      long pingDuration = System.currentTimeMillis() - pingStartTime;
```

Comment:
Same here. No need to record ping duration. i.e. you don't need to call `System.currentTimeMillis()` again here.